umbra/umbra_64.c

```diff
@@ -523,37 +501,8 @@ umbra_add_shadow_segment(umbra_map_t *map, app_segment_t *seg)
                 (seg != &app_segments[i] &&
                  segment_overlap(base, end,
                                  app_segments[i].shadow_base[map_idx],
-                                 app_segments[i].shadow_end[map_idx])) ||
-                segment_overlap(base, end,
-                                app_segments[i].reserve_base[map_idx],
-                                app_segments[i].reserve_end[map_idx])) {
+                                 app_segments[i].shadow_end[map_idx]))) {
                 ELOG(1, "ERROR: new app segment ["PFX", "PFX")'s shadow segment "
-                     "["PFX", "PFX") conflicts with app seg ["PFX", "PFX") or its "
-                     "shadow ["PFX", "PFX") or reserve ["PFX", "PFX")\n",
-                     seg->app_base, seg->app_end, base, end,
-                     app_segments[i].app_base, app_segments[i].app_end,
-                     app_segments[i].shadow_base[map_idx],
-                     app_segments[i].shadow_end[map_idx],
-                     app_segments[i].reserve_base[map_idx],
-                     app_segments[i].reserve_end[map_idx]);
-                return false;
```

Comment:
Echoing https://github.com/DynamoRIO/drmemory/issues/2283#issuecomment-684894144: This is part of the core invariant of a shadow memory scheme.  A proper shadow scheme should map any address not allocated by the application to invalid memory. Applications reference wild addresses on occasion and those should not map to application or shadow memory, causing tools to corrupt their own state or the application state. This is especially bad for tools whose goal is to detect memory bugs in applications, such as Dr. Memory.

If we were to consider relaxing this invariant, it should only be done on the mapping schemes where there is no solution otherwise, and the lack of the invariant should be made very explicit and documented so that tools that care about it (such as Dr. Memory) can avoid it.

Reply:
Thank you for raising this. So, I assumed that since UMBRA does not support meta-data for shadow regions, then checking whether a shadow's shadow collides can be relaxed because the user should never pass a shadow address for translation again. I guess a faulty app may inadvertently craft and use a shadow address due to undefined behaviour, which results in Dr Memory to translate this address again... But if the address can by anything, then why do we stop at checking whether the shadow's shadow collides with app/shadow regions, and not keep on going? i.e., check the shadow's shadow's shadow?

Reply:
> But if the address can by anything, then why do we stop at checking whether the shadow's shadow collides with app/shadow regions, and not keep on going? i.e., check the shadow's shadow's shadow and so on?

Ideally nothing would map to the shadow memory except app memory, but I think the best we can do is to ensure that either the wild address or the shadow-mapped address is invalid and will fault.  At least that's what we put in our ["Efficient Memory Shadowing for 64-bit Architectures"](http://burningcutlery.com/derek/docs/ems64-ISMM10.pdf) paper.  That's why we focused on shadow-of-shadow, b/c the app ref won't fault.  So I guess you'd live with corruption of your shadow if you have your instrumentation prior to the app load of a wild address that does map to your shadow, and then when the app load faults you have to check in a fault handler and see that it messed you up and halt the program in case the app has its own handler that continues on wild faults?  @zhaoqin do you remember whether we had some model of post-instrumentation instead of pre-instrumentation or is it how I'm describing it here?

Reply:
Okay, I understood. Thanks.

>  So I guess you'd live with corruption of your shadow if you have your instrumentation prior to the app load of a wild address that does map to your shadow, and then when the app load faults you have to check in a fault handler and see that it messed you up and halt the program in case the app has its own handler that continues on wild faults?

Yes. Basically what I am proposing is to remove reserved regions in order to make room for large scales. I know it is not ideal but haven't I thought of a better idea till now. 

>  then when the app load faults you have to check in a fault handler 

Actually, a fault might not occur at all; the app ref won't fault b/c it wildly addresses shadow memory, and the shadow's shadow address won't fault b/c it maps back to an app or shadow region as well. Hence corrupting the state

I guess I shouldn't have just removed the checks, but elided them only for larger scales as mentioned here:  https://github.com/DynamoRIO/drmemory/issues/2283#issuecomment-684895503

Reply:
> Actually, a fault might not occur at all

To clarify: I was describing the original scheme where for each app load/store address A either 1) A is in a listed app region or 2) A is invalid (will fault) or 3) shadow(A) is invalid, and how it would interact with pre-load instru.

We're on the same page: removing the shadow-shadow check breaks the invariant and now a tool can corrupt its shadow or the app without (easily) detecting it, which should be acknowledged in the interface.

Reply:
I'm closing this PR. Please see https://github.com/DynamoRIO/drmemory/pull/2309